### PR TITLE
Added ability to express operator arguments string-style

### PR DESCRIPTION
**Pull request by ben5516**
Allows this:

```ruby
def self.operators
 {between: "{date} and {date}"}
end
```

### CODE REVIEW

File: lib/set_builder/modifier/base.rb [hunk 1]
```diff
@@ -79,12 +81,20 @@ def errors_with_number_value(string)
...
+
+
+      def self.parsed_operators
+        @parsed_operators ||= begin
+          Hash[
+            operators.map do |operator, arguments|
+              [operator, Arguments.new(arguments)]
+            end
+          ]
         end
```
Comment (boblail) [inline lib/set_builder/modifier/base.rb]: :art: You don't need the `begin....end` b/c what follows (`Hash[...]`) is a single expression, so you could just remove those.

The `]` on its own line feels strange to me — _and_ the `do...end` despite the `.map` being a succinct block concerned with returning a value (not executing an instruction with side effects).

What about:

```ruby
# lisp- and java-style line-wrapping
@parsed_operators ||= Hash[operators.map { |operator, arguments|
  [operator, Arguments.new(arguments)] }]
```

or even (requiring Ruby 2.1+):

```ruby
# jquery-style chaining
@parsed_operators ||= operators
  .map { |operator, arguments| [operator, Arguments.new(arguments)] }
  .to_h
```

I think we use both those styles more frequently than:

```ruby
@parsed_operators ||= begin
  Hash[
    # ...
  ]
end
```

File: lib/set_builder/modifier/base.rb [hunk 2]
```diff
@@ -96,11 +106,11 @@ def self.negate(operator)
...
-        hash = {}
-        operators.each do |operator, array|
-          hash[operator.to_s] = array.map {|type| type.to_s }
-        end
-        hash
+        Hash[
+          parsed_operators.map do |operator, arguments|
+            [operator.to_s, arguments.as_json]
+          end
+        ]
```
Comment (boblail) [inline lib/set_builder/modifier/base.rb]: :art: same as above

File: lib/set_builder/modifier/base.rb [hunk 3]
```diff
@@ -79,12 +81,20 @@ def errors_with_number_value(string)
 
 
       def to_s(negative=false)
```
Comment (boblail) [inline lib/set_builder/modifier/base.rb]: :arrow_down: We can probably drop `negative=false` from the args now.

File: lib/set_builder/modifier/base.rb [hunk 4]
```diff
@@ -79,12 +81,20 @@ def errors_with_number_value(string)
...
-        words = negative ? [self.class.negate(operator).to_s.gsub(/_/, " ")] : [operator.to_s.gsub(/_/, " ")]
-        arguments = self.class.operators[operator] || []
-        (0...arguments.length).each do |i|
-          words << ValueMap.to_s(arguments[i], values[i])
+        arguments = self.class.parsed_operators[operator] || Arguments.new("")
```
Comment (boblail) [inline lib/set_builder/modifier/base.rb]: :arrow_down: You've got this line here and on L48. Would it make sense to extract to a method?

File: lib/assets/javascripts/set_builder.js [hunk 5]
```diff
@@ -247,7 +247,7 @@ SetBuilder.Modifiers = function(_modifiers) {
     var operator = args.operator;
     if(!operator) throw 'An operator name was not supplied.'
 
-    var params = this.params_for_operator(modifier_type, operator);
+    var params = this.params_for_operator(modifier_type, operator).__select(function(token) { return token[0] == 'arg' });
     var values = args.values;
```
Comment (boblail) [inline lib/assets/javascripts/set_builder.js]: :point_right: Do we need code to handle `toString` on the client?

File: lib/set_builder/modifier/arguments.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+          end
+        end.join
+      end
+
+    private
+
+      def map_legacy_arguments(arguments)
+        _expression = arguments.map { |arg| "{#{arg}}" }.join(" ")
+        puts "DEPRECATED: SetBuilder::Modifier should use expression style argumenents now e.g.-> \"#{_expression}\""
+        _expression
```
Comment (boblail) [inline lib/set_builder/modifier/arguments.rb]: :+1:

File: lib/set_builder/modifier/arguments.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+      def arity
+        list.count
+      end
+
+      def list
+        @types ||= @tokens
+          .select { |token, _| token == :arg }
+          .map { |_, type| type }
+      end
```
Comment (boblail) [inline lib/set_builder/modifier/arguments.rb]: You store this in the ivar `@types` — is that actually a better name for this method than `list`?

File: lib/assets/javascripts/set_builder.js (hunk 5 above)
Comment (ben5516) [inline lib/assets/javascripts/set_builder.js]: ¯\_(ツ)_/¯

**Pull request merged**